app/experimenter/experiments/constants/nimbus.py

```diff
@@ -156,25 +157,115 @@ class ApplicationConfig:
 )
 
 
+class Status(models.TextChoices):
+    DRAFT = "Draft"
+    PREVIEW = "Preview"
+    LIVE = "Live"
+    COMPLETE = "Complete"
+
+
+class PublishStatus(models.TextChoices):
+    IDLE = "Idle"
+    REVIEW = "Review"
+    APPROVED = "Approved"
+    WAITING = "Waiting"
+
+
+class LifecycleStates(Enum):
+    DRAFT_IDLE = dict(
+        status=Status.DRAFT,
+        publish_status=PublishStatus.IDLE,
+    )
+    PREVIEW_IDLE = dict(
+        status=Status.PREVIEW,
+        publish_status=PublishStatus.IDLE,
+    )
+    DRAFT_REVIEW = dict(
+        status=Status.DRAFT,
+        publish_status=PublishStatus.REVIEW,
+    )
+    DRAFT_APPROVED = dict(
+        status=Status.DRAFT,
+        publish_status=PublishStatus.APPROVED,
+    )
+    DRAFT_WAITING = dict(
+        status=Status.DRAFT,
+        publish_status=PublishStatus.WAITING,
+    )
+    LIVE_IDLE = dict(
+        status=Status.LIVE,
+        publish_status=PublishStatus.IDLE,
+    )
+    PAUSE_QUEUED = dict(
+        status=Status.LIVE,
+        is_paused=False,
+    )
+    LIVE_REVIEW_ENDING = dict(
+        status=Status.LIVE,
+        publish_status=PublishStatus.REVIEW,
+        is_end_requested=True,
+    )
+    LIVE_IDLE_REJECT_ENDING = dict(
+        status=Status.LIVE,
+        publish_status=PublishStatus.IDLE,
+        is_end_requested=False,
+    )
+    LIVE_APPROVED_ENDING = dict(
+        status=Status.LIVE,
+        publish_status=PublishStatus.APPROVED,
+        is_end_requested=True,
+    )
+    LIVE_WAITING_ENDING = dict(
+        status=Status.LIVE,
+        publish_status=PublishStatus.WAITING,
+        is_end_requested=True,
+    )
+    COMPLETE_IDLE_ENDED = dict(
+        status=Status.COMPLETE,
+        publish_status=PublishStatus.IDLE,
+        is_end_requested=True,
+    )
+
+
+class Lifecycles(Enum):
```

Comment:
First of all, INTERESTING! I hadn't thought of structuring it like this, with each entry being just the end state with the ordered tuple of preceding states.  I was thinking of it as just the list of states and then iterating over the enum leads you through the states, but that doesn't let you have multiple different branching end states like the timeouts and rejections and etc you have here, so that's a really clever design!

Second quick question, why does this have to be defined at the module level instead of inside the constants class? You can't refer to names the same way?

Reply:
> Second quick question, why does this have to be defined at the module level instead of inside the constants class? You can't refer to names the same way?

Yeah, within the scope of the constants class, the nested classes can't refer to each other.

So, without hoisting up to the module scope, LifecycleStates couldn't refer to Status or PublishStatus and Lifecycles couldn't refer to LifecycleStates